### PR TITLE
Install nodenv and check version on deploy

### PR DESCRIPTION
**Pull request by dacook**
- #838 

A new simple role was created because the ones on Ansible Galaxy didn't suit. I also tried [nodenv-installer](https://github.com/nodenv/nodenv-installer) but it didn't suit either. Sometimes it's better to build your own.

Yarn is installed automatically by a third-party nodenv package. This doesn't support specifying a yarn version like we did before, but this doesn't seem to matter (we don't specify one in openfoodnetwork codebase). If required, we could submit a PR to the package to support it.

## What should a dev test? 
This updates both the provision and deploy playbooks, so these can be tested on a local vagrant environment or uk/fr-staging. (not au-staging because I've already deployed it there).
I found my vagrant environment to be a bit faster.

### 1. Install
```
cd ofn-install
gh pr checkout 866 # e.g. using GitHub CLI which is so handy
ansible-playbook site.yml --limit=vagrant # or add secrets for staging
```

- [x] Expected: Nodenv is installed
```
TASK [fetch nodenv]
TASK [install nodenv plugins]
TASK [add nodenv to user path]
changed:
```

- [x] Expected: Node & Yarn are installed for v14.21.2
```
TASK [nodenv_install : ensure node version 14.21.2 is installed (with yarn)] ****************************************************
changed: [staging.openfoodnetwork.org.au]
Wednesday 08 March 2023  17:27:53 +1100 (0:00:05.140)       0:01:14.442 *******

TASK [nodenv_install : node installation messages] *********************************************************************
ok: [staging.openfoodnetwork.org.au] => {
    "msg": "\nWARNING: node-v14.21.2-linux-x64 is in LTS Maintenance mode and nearing its end of life.\nIt only receives *critical* security updates, *critical* bug fixes and documentation updates.\n\nInstalling node-v14.21.2-linux-x64...\nInstalled node-v14.21.2-linux-x64 to /home/openfoodnetwork/.nodenv/versions/14.21.2"
}
```

### 2. Deploy
- [x] Expected: node/yarn version is ensured based upon `.node-version` file
```
TASK [nodenv_install : ensure node version 14.21.2 is installed (with yarn)] *******************************************
ok: [local_vagrant]
```

### 3. Node upgrade
`ansible-playbook playbooks/deploy.yml --limit=vagrant -e "github_key=jibees:upgrade-node-version"`

- [ ] Expected: new version of node is installed and yarn dependencies install successfully.
```
TASK [nodenv_install : ensure node version 17.9.1 is installed (with yarn)] ********************************************
changed: [local_vagrant]
Thursday 09 March 2023  10:50:12 +1100 (0:00:18.479)       0:01:19.216 ********

TASK [nodenv_install : node installation messages] *********************************************************************
ok: [local_vagrant] => {
    "msg": "Downloading node-v17.9.1-linux-x64.tar.gz...\n-> https://nodejs.org/dist/v17.9.1/node-v17.9.1-linux-x64.tar.gz\n\nWARNING: node-v17.9.1-linux-x64 is past its end of life and is now unsupported.\nIt no longer receives bug fixes or security updates.\n\nInstalling node-v17.9.1-linux-x64...\nInstalled node-v17.9.1-linux-x64 to /home/openfoodnetwork/.nodenv/versions/17.9.1"
}
Thursday 09 March 2023  10:50:12 +1100 (0:00:00.036)       0:01:19.253 ********

TASK [deploy : run yarn install] ***************************************************************************************
changed: [local_vagrant]
Thursday 09 March 2023  10:50:25 +1100 (0:00:13.010)       0:01:32.263 ********
```

### CODE REVIEW

File: roles/deploy/tasks/deploy.yml
Comment (mkllnk) [inline roles/deploy/tasks/deploy.yml:88]: This is clever. But why don't we use the new script (other PR) for this?

**Review by mkllnk — APPROVED** [review]
Nice. great docs. Thank you.

File: roles/deploy/tasks/deploy.yml
Comment (dacook) [reply roles/deploy/tasks/deploy.yml:88]: Main reason: I got carried away when I finally got into the groove with Ansible, and forgot about the script! This does have the advantage of reporting the version number to the console.
Also I figured it would be safer to re-use the method already used here for rbenv, which never actually chdir's into the build path. I think it probably would have been fine to chdir and run the script though, and it would have been better to favour shell scripting over ansible more..

**Comment by dacook** [comment]
@mkllnk do you think it's worth doing any further testing or reviewing before merging?

**Comment by mkllnk** [comment]
> do you think it's worth doing any further testing or reviewing before merging?

A second review and independent dev test would be good. Not because I think that it would reveal any issues but because it's good for other devs to get familiar with ofn-install. :wink:

**Review by jibees — APPROVED** [review]
👌 

Thanks for this one. Ansible is still pretty mysterious to me, but this kind of PR might shed some light

**Comment by jibees** [comment]
## Testing notes

### Running `ansible-playbook site.yml --limit=vagrant`

Found:
```
TASK [node : fetch nodenv] ****************************************************************************************************************************************************************************************************
changed: [local_vagrant]
Mardi 14 mars 2023  14:46:40 +0100 (0:00:04.142)       0:13:38.376 ************ 

TASK [node : install nodenv plugins] ******************************************************************************************************************************************************************************************
changed: [local_vagrant] => (item={'name': 'node-build', 'repo': 'https://github.com/nodenv/node-build.git', 'version': 'master'})
changed: [local_vagrant] => (item={'name': 'nodenv-yarn-install', 'repo': 'https://github.com/pine/nodenv-yarn-install', 'version': 'master'})
changed: [local_vagrant] => (item={'name': 'nodenv-package-rehash', 'repo': 'https://github.com/nodenv/nodenv-package-rehash.git', 'version': 'master'})
Mardi 14 mars 2023  14:46:45 +0100 (0:00:05.631)       0:13:44.008 ************ 

TASK [node : add nodenv to user path] *****************************************************************************************************************************************************************************************
[WARNING]: Module remote_tmp /home/openfoodnetwork/.ansible/tmp did not exist and was created with a mode of 0700, this may cause issues when running as another user. To avoid this, create the remote_tmp dir with the
correct permissions manually
changed: [local_vagrant]
Mardi 14 mars 2023  14:46:46 +0100 (0:00:00.439)       0:13:44.447 ************ 

TASK [install node version (and yarn)] ****************************************************************************************************************************************************************************************
Mardi 14 mars 2023  14:46:46 +0100 (0:00:00.050)       0:13:44.497 ************ 

TASK [nodenv_install : ensure node version 14.21.2 is installed (with yarn)] **************************************************************************************************************************************************
changed: [local_vagrant]
Mardi 14 mars 2023  14:46:56 +0100 (0:00:10.009)       0:13:54.506 ************ 

TASK [nodenv_install : node installation messages] ****************************************************************************************************************************************************************************
ok: [local_vagrant] => {
    "msg": "Downloading node-v14.21.2-linux-x64.tar.gz...\n-> https://nodejs.org/dist/v14.21.2/node-v14.21.2-linux-x64.tar.gz\n\nWARNING: node-v14.21.2-linux-x64 is in LTS Maintenance mode and nearing its end of life.\nIt only receives *critical* security updates, *critical* bug fixes and documentation updates.\n\nInstalling node-v14.21.2-linux-x64...\nInstalled node-v14.21.2-linux-x64 to /home/openfoodnetwork/.nodenv/versions/14.21.2"
}
Mardi 14 mars 2023  14:46:56 +0100 (0:00:00.060)       0:13:54.567 ************ 

```

### Running `ansible-playbook playbooks/deploy.yml --limit=vagrant -e "github_key=jibees:upgrade-node-version"`

```
TASK [deploy : stop deployment if redis is both required and absent] **********************************************************************************************************************************************************
fatal: [local_vagrant]: FAILED! => {"changed": false, "msg": "The deployed OFN version requires Redis to be installed.\nPlease provision with the latest copy of ofn-install before deploying.\n"}
```
That's why I'v edited
https://github.com/openfoodfoundation/ofn-install/blob/d76e25e002ff2cdd52f036862ed5211d2304797c/roles/deploy/tasks/preflight_checks.yml#L41
to 
````
when: false
```

Then I re-run and had:

```
TASK [install node version (and yarn)] ****************************************************************************************************************************************************************************************
Mardi 14 mars 2023  14:57:22 +0100 (0:00:00.215)       0:03:12.893 ************ 

TASK [nodenv_install : ensure node version 17.9.1 is installed (with yarn)] ***************************************************************************************************************************************************
changed: [local_vagrant]
Mardi 14 mars 2023  14:57:34 +0100 (0:00:11.570)       0:03:24.464 ************ 

TASK [nodenv_install : node installation messages] ****************************************************************************************************************************************************************************
ok: [local_vagrant] => {
    "msg": "Downloading node-v17.9.1-linux-x64.tar.gz...\n-> https://nodejs.org/dist/v17.9.1/node-v17.9.1-linux-x64.tar.gz\n\nWARNING: node-v17.9.1-linux-x64 is past its end of life and is now unsupported.\nIt no longer receives bug fixes or security updates.\n\nInstalling node-v17.9.1-linux-x64...\nInstalled node-v17.9.1-linux-x64 to /home/openfoodnetwork/.nodenv/versions/17.9.1\n\nnpm notice \nnpm notice New major version of npm available! 8.11.0 -> 9.6.1\nnpm notice Changelog: <https://github.com/npm/cli/releases/tag/v9.6.1>\nnpm notice Run `npm install -g npm@9.6.1` to update!\nnpm notice "
}
Mardi 14 mars 2023  14:57:34 +0100 (0:00:00.144)       0:03:24.608 ************ 
````
and then I had a database error (`check database status`) but seems not related to this PR.


I had two issues, but not related to this PR, that's why, therefore, to me, this seems ok and I'm happy you to merge! 🔥

**Comment by dacook** [comment]
Hmm, that's a shame that your vagrant environment isn't working properly, but glad to see my changes worked at least!
Unfortunately your test didn't show if the app built successfully with the new version of node and yarn, but I think I already tested that on staging.
Besides, we will be staging the new version of node when the separate openfoodnetwork PR is tested.

So... this can be provisioned! I will start this on the staging environments and then on production.

**Pull request merged**
